File: lua/lasagna/lasagna.lua
local lasagna = {}

-- add a 'oven_time' field to the lasagna table
lasagna.oven_time = 40
local min_time = 0
local time_per_layer = 2


-- returns the remaining minutes based on the actual minutes in the oven
function lasagna.remaining_oven_time(actual_minutes_in_oven)
  return math.max(lasagna.oven_time - actual_minutes_in_oven, min_time)
end

-- calculates and returns the time needed to prepare the lasagna on the amount
-- of layers
function lasagna.preparation_time(number_of_layers)
  return time_per_layer * number_of_layers
end

-- calculates the time elapsed cooking the lasagna - including preparation time
-- and baking time.
function lasagna.elapsed_time(number_of_layers, actual_minutes_in_oven)
  return lasagna.preparation_time(number_of_layers) + actual_minutes_in_oven
end

return lasagna

File: go/lasagna/lasagna.go
package lasagna

const OvenTime int = 40

// RemainingOvenTime returns the remaining minutes based on the `actual` minutes already in the oven.
func RemainingOvenTime(actualMinutesInOven int) int {
	const minMinutesRemaining int = 0
	var remainingTime int = OvenTime - actualMinutesInOven

	if remainingTime > minMinutesRemaining {
		return remainingTime
	} else {
		return minMinutesRemaining
	}
}

// PreparationTime calculates the time needed to prepare the lasagna based on the amount of layers.
func PreparationTime(numberOfLayers int) int {
	const minutesPerLayer int = 2
	return minutesPerLayer * numberOfLayers
}

// ElapsedTime calculates the time elapsed cooking the lasagna. This time includes the preparation time and the time the lasagna is baking in the oven.
func ElapsedTime(numberOfLayers int, actualMinutesInOven int) int {
	return PreparationTime(numberOfLayers) + actualMinutesInOven
}
